### PR TITLE
fix: add better error messaging when initializing an already initialized project

### LINKED ISSUE #11579: receiving `Invalid environment name: undefined` when running `amplify init -y` in an initialized project
### How did you install the Amplify CLI?

pnpm

### If applicable, what version of Node.js are you using?

18

### Amplify CLI Version

10.5.1

### What operating system are you using?

mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

n/a

### Describe the bug

When running `amplify init -y` in an initialized project we receive `Invalid environment name: undefined`

### Expected behavior

We should receive a more helpful message such as

```
Amplify project ${appId} is already initialized for environment ${envName}. To create a new environment run `amplify add env`
```

### Reproduction steps

1. `amplify init -y`
2. `amplify init -y`

### Project Identifier

_No response_

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Additional information

_No response_

### Before submitting, please confirm:

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### PR DESCRIPTION
**Pull request by rtpascual**
#### Description of changes

Catch earlier and provide better error messaging when initializing an already initialized project with defaults (running `amplify init -y` for an already initialized project).

The error message changed when running `amplify init -y` the second time:
`Invalid environment name: undefined` -> `Amplify project <appId> is already initialized with environment <currentEnv>`

#### Issue #, if available

Fixes #11579

#### Description of how you validated changes

Added unit test for running `amplify init -y` for an initialized project and expecting the new error message.

Manually ran `amplify-dev init -y` twice, the second time gives the new error message (the same applies if initializing an app with a custom env name then running `amplify-dev init -y`):

Screenshot 2023-09-07 at 6 06 07 PM

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/__tests__/init-steps/s0-analyzeProject.test.ts [hunk 1]
```diff
@@ -75,4 +76,35 @@ describe('analyzeProject', () => {
...
+    };
+    mockContext.exeInfo.inputParams = {
+      amplify: {},
+      yes: true,
+    };
+    mockContext.usageData = {
+      emitError: emitErrorMock,
+    };
+
+    jest.spyOn(fs, 'existsSync').mockReturnValue(true);
```
Comment (rtpascual) [inline packages/amplify-cli/src/__tests__/init-steps/s0-analyzeProject.test.ts:101]: Mocked the bare minimum to replicate the situation where `amplify-dev init -y` was run the first time.

File: packages/amplify-cli/src/init-steps/s0-analyzeProject.ts [hunk 2]
```diff
@@ -220,8 +220,11 @@ const getDefaultEnv = (context: $TSContext): string | undefined => {
       resolution: INVALID_ENV_NAME_MSG,
     });
   }
-
-  if (isNewProject(context) || !context.amplify.getAllEnvs().includes(defaultEnv)) {
+  if (
+    isNewProject(context) ||
+    !context.amplify.getAllEnvs().includes(defaultEnv) ||
+    (context.exeInfo.inputParams && context.exeInfo.inputParams.yes)
```
Comment (rtpascual) [inline packages/amplify-cli/src/init-steps/s0-analyzeProject.ts:226]: This condition was added so the error message would return the default environment name of `dev` instead of `undefined`

**Review by bombguy — APPROVED** [review]

File: packages/amplify-cli/src/init-steps/s0-analyzeProject.ts [hunk 3]
```diff
@@ -241,8 +244,8 @@ const getEnvName = async (context: $TSContext): Promise<string> => {
     });
   } else if (context.exeInfo.inputParams && context.exeInfo.inputParams.yes) {
     throw new AmplifyError('ProjectInitError', {
-      message: `Invalid environment name: ${context.exeInfo.inputParams.amplify.envName}`,
-      resolution: INVALID_ENV_NAME_MSG,
+      message: `Amplify project is already initialized with environment name: ${getDefaultEnv(context)}`,
+      resolution: `To create a new environment run \`amplify add env\``,
     });
   }
```
Comment (sobolk) [inline packages/amplify-cli/src/init-steps/s0-analyzeProject.ts:250]: Couple of things:

I'm struggling to understand how do we end up in this branch if we run `amplify init -y` in already initialized project.
And therefore I'm not sure if just changing this error is safe. I.e. I'm not sure if we're not missing some other case where new error doesn't make sense. Running full e2e suite may reveal where problem could be (but it's not guaranteed to find anything). Use `yarn cloud-e2e` to trigger tests.
The caller of this code `analyzeProject` seem to set `isNewProject` flag. We could potentially use it to make sure that we throw the new error in already existing project.


Another thing is. If we want to say that `Amplify project is already initialized with environment name` we shouldn't use "default env". But rather read what env is in project files. Look for `stateManager.getCurrentEnvName` to get this value.

Apart from unit testing could you check if there's any e2e test that covers this scenario (i.e. init -y in already initialized project?). If there isn't it might be worth adding one.

Comment (rtpascual) [reply packages/amplify-cli/src/init-steps/s0-analyzeProject.ts:250]: Running `yarn cloud-e2e` didn't find anything but makes sense that we might be missing another case for this error so having the error separate and earlier would be best.

Good catch on "default env", in the case when running `amplify init` with a different environment name then `amplify init -y` we would want to get the current environment. This would also work in a multi-environment app as well.

I'll look into existing e2e tests and add one if there aren't any existing test for this scenario.

File: packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts [hunk 4]
```diff
@@ -72,4 +74,12 @@ describe('amplify init', () => {
       getParameters(projectRoot, category, authResourceName);
     }).not.toThrow();
   });
+
+  it('should fail if running amplify init -y on already initialized project', async () => {
```
Comment (sobolk) [inline packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts:78]: Please work with @bombguy and @0618 to get this included in codebuild manifests and tested in e2e test run.

**Review by sobolk — COMMENTED** [review]
LGTM.

File: packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts (hunk 4 above)
Comment (sobolk) [reply packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts:78]: Well. Skip the codebuild manifest part, since it's addition to already existing test file. But please obtain e2e run where this test passes.

File: packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts [hunk 5]
```diff
@@ -72,4 +73,11 @@ describe('amplify init', () => {
...
+    const envName = 'devtest';
+    const resourceName = 'twoInitDefaultTest';
+    await initJSProjectWithProfile(projectRoot, { disableAmplifyAppCreation: false, name: resourceName, envName });
+    await expect(initNonInteractive(projectRoot)).rejects.toThrowError('Process exited with non zero exit code 1');
```
Comment (bombguy) [inline packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts:81]: Can you not just run `initJSProjectWithProfile` twice and expect it to throw?

File: packages/amplify-e2e-core/src/init/initProjectHelper.ts [hunk 6]
```diff
@@ -516,3 +516,7 @@ export function amplifyStatus(cwd: string, expectedStatus: string, testingWithLa
 export function initHeadless(cwd: string, envName: string, appId: string): Promise<void> {
   return spawn(getCLIPath(), ['init', '--yes', '--envName', envName, '--appId', appId], { cwd, stripColors: true }).runAsync();
 }
+
+export function initNonInteractive(cwd: string): Promise<void> {
```
Comment (bombguy) [inline packages/amplify-e2e-core/src/init/initProjectHelper.ts:520]: can we just re-use the helper function ^? You should be able to just make the 2nd & 3rd parameter optional.

**Review by bombguy — DISMISSED** [review]

File: packages/amplify-e2e-core/src/init/initProjectHelper.ts (hunk 6 above)
Comment (sobolk) [reply packages/amplify-e2e-core/src/init/initProjectHelper.ts:520]: this is good feedback. let's do this.

File: packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts (hunk 5 above)
Comment (sobolk) [reply packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts:81]: `initJSProjectWithProfile` is interactive.

Comment (rtpascual) [reply packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts:81]: The `Invalid environment name: undefined` error only occurs for non-interactive init with `amplify init -y` where envName isn't passed in as an option or through the input prompt

Comment (bombguy) [reply packages/amplify-e2e-tests/src/__tests__/init-special-case.test.ts:81]: pretty weird edge case. But thanks for the clarification

File: packages/amplify-e2e-core/src/init/initProjectHelper.ts (hunk 6 above)
Comment (rtpascual) [reply packages/amplify-e2e-core/src/init/initProjectHelper.ts]: addressed this in the latest commit

**Review by bombguy — APPROVED** [review]
LGTM
